src/internal/connector/onedrive/item.go

```diff
@@ -265,10 +266,29 @@ func filterUserPermissions(perms []models.Permissionable) []UserPermission {
 			continue
 		}
 
+		entityID := ""
+		if gv2.GetUser() != nil {
+			entityID = *gv2.GetUser().GetId()
+		} else if gv2.GetGroup() != nil {
+			// TODO(meain): Add test for Groups
+			entityID = *gv2.GetGroup().GetId()
+		} else if gv2.GetApplication() != nil {
+			// TODO(meain): Applications permissions restore not
```

Comment:
I'm still trying to see how we can give access to permissions for an application. This code should ideally work, but I don't know how to do this in the UI, or if it even is possible.

Let me know if you have any idea around this.

Reply:
I'm also partly tempted to add device id as that is a possibility(because of the [datatype](https://learn.microsoft.com/en-us/graph/api/resources/sharepointidentityset?view=graph-rest-1.0)) even though I could not find any reference to being able to set permissions to device.

Reply:
hmmm I assume you didn't happen to find any documentation on how/if those fields can be set? It seems a little weird to be able to populate them for permissions and it's possible they're there just because the `sharePointIdentitySet` struct is being reused for permissions..?

Looking at the code though are we certain only one of the fields will be set?

I guess we have a few options:
1. don't persist anything but user or group IDs, collection non-catastrophic errors and/or log if there's data for the other fields. Logging/errors will at least let us know if we need to expand further. Not having (currently empty) fields won't impede future JSON deserialization as it's equivalent to having an empty value and `omitempty`
2. collect and persist all fields, rely on JSON `omitempty` tag to keep us from serializing a bunch of stuff that isn't there

First one may require more changes later on. Second one requires possible restore updated, but we could presumably roll those out at a later date if needed

Reply:
> I assume you didn't happen to find any documentation on how/if those fields can be set?

I have not been able to find anything yet. I know application is possible from the fact that it mentions the id could be from user/application in the [docs for the field](https://learn.microsoft.com/en-us/graph/api/resources/permission?view=graph-rest-1.0#properties).

>  it's possible they're there just because the sharePointIdentitySet struct is being reused for permissions

That is my best guess.

> are we certain only one of the fields will be set?

I have not found anything mentioned specifically in docs, but from the experiments that I did, it seems so. Even the docs examples, it does not have multiple set at any point.


> I guess we have a few options

Both seem like good options, though I am a bit more leaning towards the first one:
- It will let us catch and react when error happens(this comes with the problem that we will get errors)
- We are not storing unnecessary information

Reply:
I was able to find some documentation on how application level permission can be set. This seems to be only for SharePoint though, and not for OneDrive IIUC.
https://devblogs.microsoft.com/microsoft365dev/controlling-app-access-on-specific-sharepoint-site-collections/